### ISSUE TITLE
Wrong scope returned using gs.getCallerScopeName()

### ISSUE THREAD

**Issue by EricTherrien** [jneale/Xplore#54]
Consider the following piece of code:
```
var AScope = Class.create(); 
AScope.prototype = { 
    callerscope: function() { 
        return gs.getCallerScopeName(); 
    }, 
}
var obj = new AScope();
gs.debug( "[scope = " + obj.callerscope() +"]" ); 
```
If you run this within a scope from Background Script or Xplore, you get different result, which, obviously, should not be the case.

In Background Script you get the expected scope, but if you run it in Xplore, you always get `rhino.global` as the scope, no matter what scope you run your code in.

**Comment by EricTherrien** [comment]
Hi James,

Yes make plenty of senses...

About what I'm doing, it's a bit complex to explain, but I can resume it to "dynamically extending class".

It work fine when everything is in one scope (say global), but when I "mix and match" scopes, I have a need to know where the current object is to look at the right place for possible extensions to that class (because I want to extend a given class with stuff from it's scope only).

So I was exploring my options and was testing with gs.getCallerScopeName() and found the issue.

Basically, I need a way to, for a given piece of code, find in what scope it is so I can find extensions to it...

**Comment by jneale** [comment]
Hi Eric,

I totally get it - this is exactly the scenario I attempted to solve with this.

Here's a quick test/example I threw together which may help (or may not!):
```javascript
var ARoot = Class.create();
ARoot.prototype.type = 'ARoot';
ARoot.prototype.initialize = function () {
  this.caller_scope = gs.getCallerScopeName(); // does not work - is called by Script Include "Class"
}
ARoot.prototype.getRoot = function () {
  // will only work if getRoot is called directly by the third party scope
  return gs.getCallerScopeName();
};

var AChild = Class.create();
AChild.prototype = Object.extendsObject(ARoot);
AChild.prototype.type = 'AChild';
AChild.prototype.initialize = function () {
  ARoot.prototype.initialize.apply(this, arguments);
  // this.caller_scope = gs.getCallerScopeName();
}
AChild.prototype.getChild = function () {
  return this.getRoot() + '/' + gs.getCallerScopeName();
};

// must be called from the third party, or give it the third party scope directly
AChild.prototype.getAPI = function (scope) {
 	var caller = scope || gs.getCallerScopeName();
  var scope = (function () { return this;})();
  var api;
  try {
    api = scope["api_" + caller];
  } catch (e) {
    throw "API does not exist for: " + caller;
  }
  if (typeof api === 'function') {
    return new api();
  } else {
    throw "API is not a valid script include for " + caller;
  }
};

// our third party API specific to a scope
var api_sn_sc = Class.create();
api_sn_sc.prototype.initialize = function () {
	gs.print("Hello from " + this.type);
};
api_sn_sc.prototype.type = 'api_sn_sc';

// test!
var achild = new AChild();
gs.print(achild.caller_scope);
gs.print(achild.getRoot());
gs.print(achild.getChild());
achild.getAPI('sn_sc');
```
